### PR TITLE
Fix Snapshot Shard Status Request Deduplication

### PR DESCRIPTION
**Pull request by original-brownbear**
The request deduplication didn't actually work for these requests
since they had no `equals` and `hashCode` so the deduplicator wouldn't
actually recognize equal requests.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2 (known ILM test failure)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java [hunk 1]
```diff
@@ -440,6 +441,23 @@ public ShardSnapshotStatus status() {
...
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final UpdateIndexShardSnapshotStatusRequest that = (UpdateIndexShardSnapshotStatusRequest) o;
+            return snapshot.equals(that.snapshot) && shardId.equals(that.shardId) && status.equals(that.status);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: nit: I think all fields of this class can be final now. Can we also add a test using EqualsHashCodeTestUtils?

**Review by dnhatn — APPROVED** [review]
Good catch. LGTM.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java]: Thanks Nhat! How about the tests in e21440d4c5b  ? :)

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Armin for an extra iteration :).

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2 (failure is known ILM test failure)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2 (failure is known ILM test failure again ...)

**Comment by original-brownbear** [comment]
Thanks Nhat!

**Pull request merged**